src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp

```diff
@@ -359,11 +359,16 @@ void TlsCertificateManagementServer::HandleGenerateClientCsr(HandlerContext & ct
     VerifyOrReturn(req.nonce.size() == kNonceBytes, ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
 
     auto fabric = ctx.mCommandHandler.GetAccessingFabricIndex();
-    uint8_t numClientCerts;
-    VerifyOrReturn(mCertificateTable.GetClientCertificateCount(fabric, numClientCerts) == CHIP_NO_ERROR,
-                   ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure));
-    VerifyOrReturn(numClientCerts < mMaxClientCertificates,
-                   ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ResourceExhausted));
+
+    // If no CCDID is specified, ensure we have capacity for a new client certificate.
+    if (req.ccdid.IsNull())
+    {
+        uint8_t numClientCerts;
+        VerifyOrReturn(mCertificateTable.GetClientCertificateCount(fabric, numClientCerts) == CHIP_NO_ERROR,
+                       ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure));
+        VerifyOrReturn(numClientCerts < mMaxClientCertificates,
+                       ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ResourceExhausted));
+    }
 
     auto status = mDelegate.GenerateClientCsr(ctx.mRequestPath.mEndpointId, fabric, req, [&](auto & response) -> Status {
```

Comment:
Unless I am missing something, it seems the delegate function is not returning the required status codes defined in the Spec.viz., NotFound in the not-null case, DynamicConstraintError in the null case, etc. 
Not directly related to the issue this PR addresses though.
https://github.com/project-chip/connectedhomeip/blob/master/examples/all-clusters-app/all-clusters-common/src/tls-certificate-management-instance.cpp#L298.

Reply:
For the null case, I could not find the `DynamicContraintError` (or the key collission check for that error), so I'll defer to @gmarcosb if he is aware of the context.

For the non-null case, I think the `NotFound` is returned from `PrepareClientCertificate` (in `CHIP_ERROR_NOT_FOUND` if you follow that lookup trail), but it's getting reduced/generalized to `Status::Failure` in the delegate function so that's something which certainly needs to be fixed. 

As it is unrelated to the issue/fix in this PR, I can create a separate issue for these bits.

Reply:
Yes, exactly. That's what I observed that the errors from that function were converted into a Status::Failure. 
Thanks for creating the issue.